osquery/tables/events/freebsd/openbsm_events.cpp

```diff
@@ -0,0 +1,208 @@
+
+void OpenBSMExecVESubscriber::configure() {
+  std::vector<size_t> event_ids{
+      23, // execve
+      43190, // AUX_POSIX_SPAWN
+  };
+  for (const auto& evid : event_ids) {
+    auto sc = createSubscriptionContext();
+    sc->event_id = evid;
+    subscribe(&OpenBSMExecVESubscriber::Callback, sc);
+  }
+}
+
+Status OpenBSMExecVESubscriber::Callback(
+    const OpenBSMEventContextRef& ec, const OpenBSMSubscriptionContextRef& sc) {
+  Row r;
+  for (const auto& tok : ec->tokens) {
+    switch (tok.id) {
+    case AUT_HEADER32:
+      r["time"] = BIGINT(tok.tt.hdr32.s);
+      break;
+    case AUT_HEADER32_EX:
+      r["time"] = BIGINT(tok.tt.hdr32_ex.s);
+      break;
+    case AUT_HEADER64:
+    case AUT_HEADER64_EX:
+      r["time"] = BIGINT(tok.tt.hdr64_ex.s);
+      break;
+    case AUT_SUBJECT32:
+      r["auid"] = INTEGER(tok.tt.subj32.auid);
+      r["pid"] = INTEGER(tok.tt.subj32.pid);
+      r["euid"] = INTEGER(tok.tt.subj32.euid);
+      r["egid"] = INTEGER(tok.tt.subj32.egid);
+      r["uid"] = INTEGER(tok.tt.subj32.ruid);
+      r["gid"] = INTEGER(tok.tt.subj32.rgid);
+      break;
+    case AUT_SUBJECT64:
+      r["auid"] = INTEGER(tok.tt.subj64.auid);
+      r["pid"] = INTEGER(tok.tt.subj64.pid);
+      r["euid"] = INTEGER(tok.tt.subj64.euid);
+      r["egid"] = INTEGER(tok.tt.subj64.egid);
+      r["uid"] = INTEGER(tok.tt.subj64.ruid);
+      r["gid"] = INTEGER(tok.tt.subj64.rgid);
+      break;
+    case AUT_SUBJECT32_EX:
+      OpenBSM_AUT_SUBJECT32_EX(r, tok);
+      break;
+    case AUT_RETURN32:
+      r["status"] = INTEGER(static_cast<unsigned long>(tok.tt.ret32.status));
+      break;
+    case AUT_RETURN64:
+      r["status"] = INTEGER(tok.tt.ret64.err);
+      break;
+    case AUT_EXEC_ARGS:
+      for (size_t i = 0; i < tok.tt.execarg.count; ++i) {
+        r["cmdline"] += std::string(tok.tt.execarg.text[i]) + " ";
+      }
+      r["cmdline_size"] = INTEGER(r["cmdline"].length());
+      break;
+    case AUT_PATH:
+      r["path"] = std::string(tok.tt.path.path);
+      break;
+    case AUT_ATTR32:
+      r["mode"] = INTEGER(decimalIntToOctInt(tok.tt.attr32.mode));
+      r["owner_uid"] = INTEGER(tok.tt.attr32.uid);
+      r["owner_gid"] = INTEGER(tok.tt.attr32.gid);
+      r["fsid"] = INTEGER(tok.tt.attr32.fsid);
+      r["nid"] = INTEGER(tok.tt.attr32.nid);
+      r["dev"] = INTEGER(tok.tt.attr32.dev);
+      break;
+    case AUT_EXEC_ENV:
+      for (size_t i = 0; i < tok.tt.execarg.count; ++i) {
+        r["env"] += std::string(tok.tt.execenv.text[i]) + " ";
+      }
+      r["env_count"] = INTEGER(tok.tt.execarg.count);
+      r["env_size"] = INTEGER(r["env"].length());
+      break;
+    }
+  }
+  r["uptime"] = INTEGER(tables::getUptime());
+  add(r);
+  return Status(0, "OK");
+}
+
+void OpenBSMSSHLoginSubscriber::configure() {
+  auto sc = createSubscriptionContext();
+  sc->event_id = 32800;
+  subscribe(&OpenBSMSSHLoginSubscriber::Callback, sc);
+}
+
+Status OpenBSMSSHLoginSubscriber::Callback(
```

Comment:
Does this work on FreeBSD?

Reply:
I don't see anything platform-dependent here — was there something that jumped out at you?

Reply:
This code that this listens too might not fire on FreeBSD

Reply:
Yeah, should fire just fine, the event is there on all platforms.